sbtc-core/src/operations/commit_reveal/construction.rs

```diff
@@ -0,0 +1,157 @@
+    Opcode,
+};
+
+/// Data to construct a commit reveal deposit transaction
+pub struct DepositData<'p> {
+    /// Address to deposit to
+    pub address: &'p StacksAddress,
+    /// Name of the contract to deposit to
+    pub contract_name: Option<&'p str>,
+    /// How much to send for the reveal fee
+    pub reveal_fee: u64,
+}
+
+impl<'p> DepositData<'p> {
+    /// Create deposit data
+    pub fn new(
+        address: &'p StacksAddress,
+        contract_name: Option<&'p str>,
+        reveal_fee: u64,
+    ) -> Self {
+        Self {
+            address,
+            contract_name,
+            reveal_fee,
+        }
+    }
+
+    /// Serializes this data according to the SIP-021 wire formats
+    /// Links:
+    ///  - [SIP draft](https://github.com/stacksgov/sips/blob/56b73eada5ef1b72376f4a230949297b3edcc562/sips/sip-021/sip-021-trustless-two-way-peg-for-bitcoin.md)
+    ///  - [Reference implementation](https://github.com/stacks-network/stacks-blockchain/blob/next/src/chainstate/burn/operations/peg_in.rs)
+    pub fn to_vec(&self) -> Vec<u8> {
+        once(Opcode::Deposit as u8)
+            .chain(once(self.address.version() as u8))
+            .chain(self.address.hash().as_ref().iter().cloned())
+            .chain(
+                self.contract_name
+                    .map(|contract_name| contract_name.as_bytes().to_vec())
+                    .into_iter()
+                    .flatten(),
+            )
+            .chain(repeat(0))
+            .take(78)
+            .chain(self.reveal_fee.to_be_bytes())
+            .collect()
+    }
+}
+
+/// Data to construct a commit reveal withdrawal transaction
+pub struct WithdrawalData<'r> {
+    /// Amount to withdraw
+    pub amount: u64,
+    /// Signature of the transaction
+    pub signature: &'r RecoverableSignature,
+    /// How much to send for the reveal fee
+    pub reveal_fee: u64,
+}
+
+impl<'r> WithdrawalData<'r> {
+    /// Create withdrawal data
+    pub fn new(amount: u64, signature: &'r RecoverableSignature, reveal_fee: u64) -> Self {
+        Self {
+            amount,
+            signature,
+            reveal_fee,
+        }
+    }
+
+    /// Serialize withdrawal data
+    pub fn to_vec(&self) -> CommitRevealResult<Vec<u8>> {
+        let (recovery_id, signature_bytes) = self.signature.serialize_compact();
+        let recovery_id: u8 = recovery_id
+            .to_i32()
+            .try_into()
+            .map_err(CommitRevealError::InvalidRecoveryId)?;
+        let empty_memo = [0; 4];
+
+        Ok(once(Opcode::WithdrawalRequest as u8)
+            .chain(self.amount.to_be_bytes())
+            .chain(once(recovery_id))
+            .chain(signature_bytes)
+            .chain(empty_memo)
+            .chain(self.reveal_fee.to_be_bytes())
+            .collect())
+    }
+}
+
+/// Constructs a peg in payment address
+pub fn deposit_commit<'p>(
+    deposit_data: DepositData<'p>,
+    revealer_key: &XOnlyPublicKey,
```

Comment:
Does this revealer key refer to the revealer service or the signer who actually constructs the corresponding reveal BTC transaction? I assume the signer. which means that the user could dictate which signer triggers the reveal transaction ? This implies technically that anyone could construct a BTC transaction and sign it themselves and the signers potentially need to do more than just reveal transactions. They need to check for otehr revealed transactions that some other entity has revealed so they can validate it construct the corresponding stx transaction.

Reply:
In mini, this revealer_key would be the aggregate public key from the DKG rounds corresponding private key then?

Reply:
@jferrant and I discussed this and related issues today.  We came to a number of conclusions which we will want to discuss amongst the team tomorrow.  I'll put some of them down here for reference:

1. For commit/reveal deposits, the commit txo should be a taproot txo with two spend paths: a key spend using the sbtc wallet key, and a script spend so the sender can reclaim
2. The signers only need to know the merkle root of the script tree to be able to do a key spend, which minimizes what needs to be backchanneled
3. Commit txos can be trivially stolen if spendable by an external reveal service; the sbtc wallet however has a threshold signature with the reputation of the sbtc system to prevent fraud
4. For OP_RETURN deposits, if the deposit txo is a taproot txo then the OP_RETURN txo needs to communicate the merkle root in addition to the STX principal
5. OP_RETURN/taproot deposits should be spent during the signing process if the deposit is accepted to prevent reclaim after mint
6. OP_RETURN/p2pkh deposits don't need to be spent if deposit is accepted, but then sender will have no way to reclaim if deposit is rejected

Open questions:

1. If an OP_RETURN/p2pkh deposit is rejected, what is the best course of action legally?  Is it okay to return a deposit to a (possibly) prohibited individual?  Would it be better to burn the txo?  Is there a FINCEN/OFAC/etc place to surrender such txos?
2. If we can't return and/or shouldn't burn a rejected deposit, what do we do on wallet handoff?  Do we need to leave any such utxos in the old sbtc wallet forever?  

@netrome let's discuss this tomorrow

Reply:
Happy to discuss commit-reveal. Regarding some points:

2. The signers also need to know the internal public key to spend the script path, and the merkle path to the script.
3. Yes, committers should only work with revealers they trust. This was also noted in the original design doc:

> Security note: The Revealer has the possibility to steal the commit transaction for it's own use. Doing this is a visible crime on the chain, but it relies on the Revealer to be a trusted entity. If the Revealer is operated by Stackers, it is possible to implement mechanisms to penalize such behavior.

https://github.com/Trust-Machines/stacks-sbtc/blob/main/sbtc-planning/commit-reveal-ops.md#commit-reveal-protocol

4. I'm a bit confused here. Let's talk more.

5. I assume you mean OP_DROP?

6. Yes, that's true. That's why we don't support OP_RETURN for 0.1.

**Open questions**
1. I also wonder about this one. A related question, what happens if you just receive a random payment from a blacklisted address? Should you return it or burn it?

2. Agree, we need to discuss this. This could potentially require sBTC UX changes such as dropping OP_RETURN support for 1.0 - which I have learned some people are assuming that we do.

Reply:
I sent and invite to explore this before our team meeting today. Let me know if the time doesn't work.

Reply:
> In mini, this revealer_key would be the aggregate public key from the DKG round?

Yes